### ISSUE TITLE
[Ubuntu 20.04]: [20.10.23 & 23.0.1] Docker socket not cleaned up after uninstall 

### ISSUE THREAD

**Issue by iusetabs** [moby/moby#45059]
### Description

# TL;DR

* Install docker version 20.10.23 and uninstall docker then check the docker.socket service
* Install docker version 19.03.15 and uninstall docker then check the docker.socket service
* Notice the docker.socket is not found after uninstalling 19.03.15, but is present after uninstalling 20.10.23 and 23.0.1
* docker.socket service after uninstalling versions 20+
 ```
$ systemctl status docker.socket
● docker.socket
     Loaded: not-found (Reason: Unit docker.socket not found.)
     Active: active (listening) since XXXXXXXX; 2min 21s ago
     CGroup: /system.slice/docker.socket
```
* docker.socket service after uninstalling version 19 
```
$ systemctl status docker.socket
Unit docker.socket could not be found.
```
# What

* For some testing purposes I have to work with docker versions 19.03.15, 20.10.23 and 23.0.1 on the same Ubuntu 20.04 host. 
* I have automated the docker installation process and recently I have been getting failures as when installing `docker-ce`, it fails to start. 
* The error looks like the below: 
```
$ journalctl -r --unit=docker.service --no-pager -n 100
ip-X-X-X-XXX systemd[1]: docker.service: Failed with result 'exit-code'.
ip-X-X-X-XXX systemd[1]: docker.service: Start request repeated too quickly.
ip-X-X-X-XXX  systemd[1]: Stopped Docker Application Container Engine.
ip-X-X-X-XXX  systemd[1]: docker.service: Scheduled restart job, restart counter is at 4.
ip-X-X-X-XXX  systemd[1]: Failed to start Docker Application Container Engine.
ip-X-X-X-XXX  systemd[1]: docker.service: Failed with result 'exit-code'.
ip-X-X-X-XXX  systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
ip-X-X-X-XXX dockerd[41808]: failed to load listeners: no sockets found via socket activation: make sure the service was started by systemd
```
* Immediately restarting the docker service will fail
* If you try to install docker-ce immediately after the first attempt to install docker-ce fails, it may succeed. 
* The docker service eventually restarts successfully after 30 seconds - to a minute. ( You need to restart the service manually )
* In my case, I did not want to write boilerplate code which automates this testing to add sleep / re-install attempts etc. for this specific case. 
* I figured that for versions 20.10.23 and 23.0.1 that when I uninstall them, I must additionally stop the docker.socket service. 
* This issue does then not occur if the docker.socket service is stopped when installing a new version of docker-ce.
* Also rebooting the system after uninstalling docker will also prevent this issue. I assume this removes the zombie docker.socket service. 

# Workaround

* When installing a new version of docker-ce on Ubuntu 20.04, ensure that docker.socket service is stopped ( if it exists ). 
`$ systemctl stop docker.socket`
* Applies when trying to install docker-ce versions 20.10.23 or 23.0.1 after docker-ce versions 20.10.23 or 23.0.1 have been previously installed on the system and no reboot has been carried out.

# Potential fix

* I guess there is some code which removes this docker.socket service missing from the uninstallation scripts for these docker versions? 

### Reproduce

1. `sudo apt-get remove docker docker-engine docker.io containerd runc`
2. `sudo apt-get update --assume-yes`
3. `sudo apt-get install -assume-yes 
    apt-transport-https 
    ca-certificates 
    curl 
    gnupg-agent 
    software-properties-common`
4. `curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add -`
5. `sudo add-apt-repository 
   "deb [arch=amd64] https://download.docker.com/linux/ubuntu 
   $(lsb_release -cs) 
   stable"`
6. `sudo apt-get  update  --assume-yes`
7. `sudo apt-get install --assume-yes docker-ce="5:20.10.23~3-0~ubuntu-focal" docker-ce-cli="5:20.10.23~3-0~ubuntu-focal" containerd.io="1.6.10-1"`
8.  `sudo apt-get purge --assume-yes docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin docker-ce-rootless-extras`
9.  `sudo rm -rf /var/lib/docker`
10.  `sudo rm -rf /var/lib/containerd`
11. `sudo apt-get install --assume-yes docker-ce="5:20.10.23~3-0~ubuntu-focal" docker-ce-cli="5:20.10.23~3-0~ubuntu-focal" containerd.io="1.6.10-1"`
// here is the point of failure

### Expected behavior

The docker socket should be removed upon removal of docker-ce. 

### docker version

```bash
Client: Docker Engine - Community
 Version:           20.10.23
 API version:       1.41
 Go version:        go1.18.10
 Git commit:        7155243
 Built:             Thu Jan 19 17:36:25 2023
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.23
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.10
  Git commit:       6051f14
  Built:            Thu Jan 19 17:34:14 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.10
  GitCommit:        770bd0108c32f3fb5c73ae1264f7e503fe7b2661
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Is this needed?
```

### Additional Info

This issue was happening on my Ubuntu 20.04 AWS EC-2 instances. But assuming AWS is not related to this issue.

**Comment by neersighted** [comment]
Closing as this was solved in docker-ce-packaging, and packaging is not in scope for moby/moby. Please open an issue on docker/docker-ce-packaging if you think this is not solved with the latest builds.
